nautobot/core/management/commands/pre_migrate.py

```diff
@@ -120,6 +120,48 @@ def check_configcontext_uniqueness():
         )
 
 
+def check_interface_ipaddress_vrf_uniqueness(command):
+    """
+    Check for uniqueness enforcement changes for Interface and VMInterface IPAddress assignments.
+
+    - All IP Addresses assigned to a specific Interface or VMInterface must share the same VRF or have no VRF.
+    """
+
+    failed = False
+
+    command.stdout.write(command.style.WARNING(">>> Running interface/vminterface vrf uniqueness checks..."))
+
+    for ip_address_assignment in (
+        IPAddress.objects.values("assigned_object_type", "assigned_object_id")
+        .annotate(
+            vrf_count=models.Count("vrf_id", distinct=True),
+            vrf_null_count=models.Count("pk", filter=models.Q(vrf_id__isnull=True)),
+        )
+        .order_by()
+    ):
+        has_null_vrf = 1 if ip_address_assignment["vrf_null_count"] > 0 else 0
+        if (
+            ip_address_assignment["vrf_count"] + has_null_vrf > 1
+            and ip_address_assignment["assigned_object_id"] is not None
+        ):
+            failed = True
+            assigned_object_type = ContentType.objects.get_for_id(
+                ip_address_assignment["assigned_object_type"]
+            ).model_class()
+            assigned_object = assigned_object_type.objects.get(pk=ip_address_assignment["assigned_object_id"])
```

Comment:
Also should account for possibility of DoesNotExist exception being thrown here.